test/unit/error.spec.ts

```diff
@@ -292,3 +285,94 @@ test("return from nested finally", () => {
     `;
     expect(util.transpileAndExecute(code)).toBe("finally AB");
 });
+
+test.each([
+    `"error string"`,
+    `42`,
+    `3.141`,
+    `true`,
+    `false`,
+    `undefined`,
+    `{ x: "error object" }`,
+    `() => "error function"`,
+])("throw and catch arbitrary types", error => {
+    util.testFunction`
+            try {
+                throw ${error};
+            } catch (error) {
+                if (typeof error == 'function') {
+                    return error();
+                } else {
+                    return error;
+                }
+            }
+        `.expectToMatchJsResult();
+});
+
+test.each([
+    "Error",
+    "RangeError",
+    "ReferenceError",
+    "SyntaxError",
+    "TypeError",
+    "URIError",
+    "new Error",
+    "new RangeError",
+    "new ReferenceError",
+    "new SyntaxError",
+    "new TypeError",
+    "new URIError",
+])("test builtin error properties", errorType => {
+    util.testFunction`
+        const error = ${errorType}();
+        return error.name;
+    `.expectToMatchJsResult();
+
+    util.testFunction`
+        const error = ${errorType}();
+        return error.message;
+    `.expectToMatchJsResult();
+
+    util.testFunction`
+        const error = ${errorType}();
+        return error.toString();
+    `.expectToMatchJsResult();
+});
+
+test.each([
+    "Error",
+    "RangeError",
+    "ReferenceError",
+    "SyntaxError",
+    "TypeError",
+    "URIError",
+    "new Error",
+    "new RangeError",
+    "new ReferenceError",
+    "new SyntaxError",
+    "new TypeError",
+    "new URIError",
+    "new CustomError",
+])("get stack from %s", errorType => {
+    const stack = util.testFunction`
+        class CustomError extends Error {
+            public name = "CustomError";
+        }
+
+        function innerFunctionThatThrows() { throw ${errorType}(); }
+        function outerFunctionThatThrows() { innerFunctionThatThrows(); }
+
+        try {
+            outerFunctionThatThrows();
```

Comment:
Stack catching is also not really related to throwing, so we can just return error object from these test functions

Reply:
Due to Lua's tail call optimisation, it actually won't work for simply returning the error object (from Lua's perspective the stack is at the outer call). We could work around this by storing the object in temporary variables, but I'd prefer to just keep the exception stack check.